### PR TITLE
Feat: Housing

### PR DESCRIPTION
**Pull request by AngeloTadeucci**
Since this PR is HUGE I'm looking for any kind of feedback, code or gameplay wise.

- #73 

### Main files:
- Home:
    - Entity saved to database with all home properties;
    - Simple level up and experience system for decoration score.
- HomeTemplates:
    -  Contains all blocks for the templates.
- HomeLayouts:
    - Entity to save home layouts to database.
- HomeManager:
    - Holds all homes on memory so multiple players can access home data at the same time;
    - Also caches which map plots has been loaded from database already;
    - Assigns two instance to each home, one for the normal plot and one for the decor planner;
    - When instantiated, creates an Task to update all homes; 
    - **_Maybe adding sometype of boolean check if home has been edited before trying to save to DB?_**.
- RequestCubeHandler:
    - Implemented methods for: name, description, password, permissions, increase/decrease size/height, change background/lighting/camera, building rights and others
- RequestHomeHandler:
    - Implemented invite other players to home, new home creation and teleporting to home.
- Player/FieldManager:
    - Fixed **Warp** method: it wasn't updating coords if rotation was default or (0, 0, 0);
    - Changed `InstanceId` type to **long** so homes can have it own instance based on it's ID;
    - Loading `IFieldObject<Cube>` for each home from Player.VisitingHomeId when the first player enters the field, sadly **FieldManager** isn't instantiated when loading home cubes in `LoadUGCMapHandler`.
- LoadUgcMapHandler:
    - Loads each home properties and sends all cubes to client;
    - If map isn't _PrivateResidence_ load all home plots for that map id;

### Important packet files:
- **ResponseCubePacket**
- **ResponseLoadUGCMapPacket**
- **SendCubesPacket**
- HomeCommandPacket
- InviteToHomePacket
- EnterUGCMapPacket
- FieldObjectPacket
- FieldPacket
- FurnishingInventoryPacket
- WarehouseInventoryPacket
- GuildPacket

### Missing:
- Life skills blocks;
- Portal cubes;
- Maids

### CODE REVIEW

**Review by Sparkymod — COMMENTED** [review]
Looks good

File: MapleServer2/Packets/ResponseLoadUGCMapPacket.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+                pWriter.WriteInt(home.ArchitectScoreTotal);
+                pWriter.WriteInt();
+                pWriter.WriteInt();
+                pWriter.WriteByte();
+                pWriter.WriteByte(home.Size);
+                pWriter.WriteByte(home.Height);
+                pWriter.WriteByte(home.Background);
+                pWriter.WriteByte(home.Lighting);
+                pWriter.WriteByte(home.Camera);
+                pWriter.WriteByte(9);
```
Comment (AlanMorel) [inline MapleServer2/Packets/ResponseLoadUGCMapPacket.cs:32]: Does `home.Permissions.Length/Size/Count` not work here?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Packets/ResponseLoadUGCMapPacket.cs (hunk 1 above)
Comment (AngeloTadeucci) [reply MapleServer2/Packets/ResponseLoadUGCMapPacket.cs:32]: Oh yes I could do that, oops!

Comment (AngeloTadeucci) [reply MapleServer2/Packets/ResponseLoadUGCMapPacket.cs:32]: No I can't, permissions don't always have all entries!

Comment (AlanMorel) [reply MapleServer2/Packets/ResponseLoadUGCMapPacket.cs:32]: Oh yeah true. Any way for 9 to not be a magic number? Perhaps a static constant inside the `Home` file?

Comment (AlanMorel) [reply MapleServer2/Packets/ResponseLoadUGCMapPacket.cs:32]: Unrelated, but what about ripping out lines 18 to 44 into a separate method called `WriteHome(home)`, I feel like this probably isn't the only time we'll need this, and it looks nicer. Optional though

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/PacketHandlers/Game/MoveFieldHandler.cs [hunk 2]
```diff
@@ -140,13 +153,90 @@ public static void HandleMove(GameSession session, PacketReader packet)
...
+            returnCoord.Z += Block.BLOCK_SIZE;
+            player.Warp(player.ReturnMapId, returnCoord, player.Rotation);
+            player.ReturnMapId = 0;
+            player.VisitingHomeId = 0;
+        }
+
+        private static void HandleEnterDecorPlaner(GameSession session)
+        {
+            Player player = session.Player;
+            if (!player.IsInDecorPlanner)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/MoveFieldHandler.cs:231]: Invert and return

**Review by AlanMorel — CHANGES_REQUESTED** [review]

**Review by AlanMorel — APPROVED** [review]

File: GameDataParser/Parsers/ItemParser.cs [hunk 3]
```diff
@@ -499,6 +500,20 @@ protected override List<ItemMetadata> Parse()
 
                 metadata.Gender = byte.Parse(limit.Attributes["genderLimit"].Value);
 
+                XmlNode install = item.SelectSingleNode("install");
+                bool isCubeSolid = byte.Parse(install.Attributes["cubeProp"].Value) == 1;
+                metadata.IsCubeSolid = isCubeSolid;
+
+                XmlNode housing = item.SelectSingleNode("housing");
```
Comment (Clockworkx) [inline GameDataParser/Parsers/ItemParser.cs]: housingNode and housing variable

File: MapleServer2/Data/Static/UGCMapMetadataStorage.cs [hunk 4]
```diff
@@ -26,11 +26,9 @@ public static bool IsValid(int mapId)
             return map.ContainsKey(mapId);
         }
 
-        public static UGCMapGroup GetMetadata(int mapId, byte groupId)
-        {
-            UGCMapMetadata mapMetadata = map.GetValueOrDefault(mapId);
-            return mapMetadata.Groups.FirstOrDefault(x => x.Id == groupId);
-        }
+        public static UGCMapGroup GetGroupMetadata(int mapId, byte groupId) => GetMetadata(mapId).Groups.FirstOrDefault(x => x.Id == groupId);
```
Comment (Clockworkx) [inline MapleServer2/Data/Static/UGCMapMetadataStorage.cs:29]: formatting?

File: MapleServer2/Database/DatabaseContext.cs [hunk 5]
```diff
@@ -43,8 +44,8 @@ public class DatabaseContext : DbContext
         public DbSet<UGCMapExtensionSaleEvent> Event_UGCMapExtensionSale { get; set; }
         public DbSet<FieldPopupEvent> Event_FieldPopup { get; set; }
         public DbSet<CardReverseGame> CardReverseGame { get; set; }
-        // public DbSet<Home> Homes { get; set; }
-
+        public DbSet<Home> Homes { get; set; }
+        public DbSet<HomeLayout> Home_Layouts { get; set; }
```
Comment (Clockworkx) [inline MapleServer2/Database/DatabaseContext.cs:48]: underscore syntax?

File: MapleServer2/Data/Static/UGCMapMetadataStorage.cs (hunk 4 above)
Comment (AngeloTadeucci) [reply MapleServer2/Data/Static/UGCMapMetadataStorage.cs:29]: what's the issue?

File: MapleServer2/PacketHandlers/Game/BuildModeHandler.cs [hunk 6]
```diff
@@ -40,14 +40,23 @@ public override void Handle(GameSession session, PacketReader packet)
 
         private static void HandleStop(GameSession session)
         {
+            if (session.FieldPlayer.Value.Guide == null)
+            {
+                return;
+            }
             session.Send(BuildModePacket.Use(session.FieldPlayer, false));
-            session.FieldManager.BroadcastPacket(GuideObjectPacket.Remove(session.Player.Guide));
+            session.FieldManager.BroadcastPacket(GuideObjectPacket.Remove(session.FieldPlayer.Value.Guide));
```
Comment (Clockworkx) [inline MapleServer2/PacketHandlers/Game/BuildModeHandler.cs:48]: formatting

File: MapleServer2/PacketHandlers/Game/DungeonHandler.cs [hunk 7]
```diff
@@ -63,7 +63,7 @@ public override void Handle(GameSession session, PacketReader packet)
 
         public static void HandleEnterDungeonPortal(GameSession session)
         {
-            int instanceId = session.Player.InstanceId;
+            long instanceId = session.Player.InstanceId;
             DungeonSession dungeonSession = GameServer.DungeonManager.GetDungeonSessionByInstanceId(instanceId);
```
Comment (Clockworkx) [inline MapleServer2/PacketHandlers/Game/DungeonHandler.cs:67]: formatting wrong or is that a github bug o.O

File: MapleServer2/PacketHandlers/Game/LoadUgcMapHandler.cs [hunk 8]
```diff
@@ -13,11 +21,77 @@ public class LoadUgcMapHandler : GamePacketHandler
...
+                Home home = GameServer.HomeManager.GetHome(session.Player.VisitingHomeId);
+                if (home == null)
+                {
+                    session.Send(ResponseLoadUGCMapPacket.LoadUGCMap(false));
+                    return;
+                }
+
+                homes = new List<Home>() { home };
+
+                session.Send(ResponseLoadUGCMapPacket.LoadUGCMap(mapIsHome, home, session.Player.IsInDecorPlanner));
```
Comment (Clockworkx) [inline MapleServer2/PacketHandlers/Game/LoadUgcMapHandler.cs:44]: formatting

File: MapleServer2/PacketHandlers/Game/LoadUgcMapHandler.cs [hunk 9]
```diff
@@ -13,11 +21,77 @@ public class LoadUgcMapHandler : GamePacketHandler
...
+
+                // Find spawning coords for home
+                int cubePortalId = 50400190;
+                List<Cube> portals = home.FurnishingInventory.Values.Where(x => x.Item != null && x.Item.Id == cubePortalId).ToList();
```
Comment (Clockworkx) [inline MapleServer2/PacketHandlers/Game/LoadUgcMapHandler.cs:48]: formatting

Comment (Clockworkx) [reply MapleServer2/PacketHandlers/Game/LoadUgcMapHandler.cs:48]: formatting

**Review by Sparkymod — APPROVED** [review]

**Review by Clockworkx — APPROVED** [review]

**Review by Clockworkx — CHANGES_REQUESTED** [review]
On the test server there were some unhandled exceptions related to housing. Tade knows about it, perhaps makes sure those are fixed
It had to do with a new account choosing a template and his house the first time. If it's not reproducable it does not matter.
Exception Type: System.NullReferenceException
Message: Object reference not set to an instance of an object.
Stack Trace:    at MapleServer2.Packets.ResponseLoadUGCMapPacket.LoadUGCMap(Boolean isHome, Home home, Boolean inDecorPlanner) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Packets\ResponseLoadUGCMapPacket.cs:line 48
   at MapleServer2.PacketHandlers.Game.LoadUgcMapHandler.Handle(GameSession session, PacketReader packet) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\PacketHandlers\Game\LoadUgcMapHandler.cs:line 44
   at MapleServer2.Network.PacketRouter1.OnPacket(Object sender, Packet packet) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Network\PacketRouter.cs:line 29
   at MapleServer2.Network.Session.StartRead() in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Network\Session.cs:line 249
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_1(Object state) in System.Private.CoreLib.dll:token 0x6002e89+0x0
   at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi) in System.Private.CoreLib.dll:token 0x6002b9e+0xe
   at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action1 callback, TState& state) in System.Private.CoreLib.dll:token 0x6002a31+0x1a
   at System.Threading.QueueUserWorkItemCallback.Execute() in System.Private.CoreLib.dll:token 0x6002b9a+0x6
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002b7c+0x110
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002c66+0x67
   at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600280f+0xe

**Comment by AngeloTadeucci** [comment]
> On the test server there were some unhandled exceptions related to housing. Tade knows about it, perhaps makes sure those are fixed
> It had to do with a new account choosing a template and his house the first time. If it's not reproducable it does not matter.
> Exception Type: System.NullReferenceException
> Message: Object reference not set to an instance of an object.
> Stack Trace: at MapleServer2.Packets.ResponseLoadUGCMapPacket.LoadUGCMap(Boolean isHome, Home home, Boolean inDecorPlanner) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Packets\ResponseLoadUGCMapPacket.cs:line 48
> at MapleServer2.PacketHandlers.Game.LoadUgcMapHandler.Handle(GameSession session, PacketReader packet) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\PacketHandlers\Game\LoadUgcMapHandler.cs:line 44
> at MapleServer2.Network.PacketRouter1.OnPacket(Object sender, Packet packet) in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Network\PacketRouter.cs:line 29
> at MapleServer2.Network.Session.StartRead() in C:\Users\Administrator\Desktop\MapleServer2\mapleserver2\Network\Session.cs:line 249
> at System.Threading.Tasks.Task.<>c.b__140_1(Object state) in System.Private.CoreLib.dll:token 0x6002e89+0x0
> at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi) in System.Private.CoreLib.dll:token 0x6002b9e+0xe
> at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action1 callback, TState& state) in System.Private.CoreLib.dll:token 0x6002a31+0x1a
> at System.Threading.QueueUserWorkItemCallback.Execute() in System.Private.CoreLib.dll:token 0x6002b9a+0x6
> at System.Threading.ThreadPoolWorkQueue.Dispatch() in System.Private.CoreLib.dll:token 0x6002b7c+0x110
> at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in System.Private.CoreLib.dll:token 0x6002c66+0x67
> at System.Threading.Thread.StartCallback() in System.Private.CoreLib.dll:token 0x600280f+0xe

Fixed!

**Review by Clockworkx — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

**Review by Sparkymod — APPROVED** [review]

**Pull request merged**